### PR TITLE
Make ClusterControl.membershipSnapshot async property

### PR DESCRIPTION
**Pull request by yim-lee**
Resolves https://github.com/apple/swift-distributed-actors/issues/949

### CODE REVIEW

File: Sources/DistributedActors/Cluster/ClusterControl.swift [hunk 1]
```diff
@@ -42,37 +43,62 @@ public struct ClusterControl {
...
+        get async throws {
+            try await self._membershipSnapshotHolder.get()
+        }
     }
 
     internal func updateMembershipSnapshot(_ snapshot: Cluster.Membership) {
-        self.membershipSnapshotLock.lock()
-        defer { self.membershipSnapshotLock.unlock() }
-        self._membershipSnapshotHolder.membership = snapshot
+        Task {
```
Comment (yim-lee) [inline Sources/DistributedActors/Cluster/ClusterControl.swift:52]: Not sure about usages of `Task`.

File: Samples/Sources/SampleDiningPhilosophers/DistributedDiningPhilosophers.swift [hunk 2]
```diff
@@ -76,4 +72,14 @@ final class DistributedDiningPhilosophers {
...
     }
+
+    private func isClusterFormed(_ systems: [ClusterSystem]) async throws -> Bool {
+        for system in systems {
+            let upCount = try await system.cluster.membershipSnapshot.count(withStatus: .up)
+            if upCount != systems.count {
+                return false
+            }
+        }
+        return true
```
Comment (ktoso) [inline Samples/Sources/SampleDiningPhilosophers/DistributedDiningPhilosophers.swift:83]: Implementing this but with a continuation and condition like "`awaitMembers(count: ..., statusAtLeast: .up)`" etc is actually ticket https://github.com/apple/swift-distributed-actors/issues/948 :) 

It is a less urgent one, but FYI, maybe a TODO here?

File: Sources/DistributedActors/ClusterSystem.swift [hunk 3]
```diff
@@ -491,8 +497,10 @@ public class ClusterSystem: DistributedActorSystem, @unchecked Sendable {
         self.shutdownSemaphore.wait()
 
         /// Down this member as part of shutting down; it may have enough time to notify other nodes on an best effort basis.
-        if let myselfMember = self.cluster.membershipSnapshot.uniqueMember(self.cluster.uniqueNode) {
-            self.cluster.down(member: myselfMember)
+        Task {
+            if let myselfMember = try await self.cluster.membershipSnapshot.uniqueMember(self.cluster.uniqueNode) {
+                self.cluster.down(member: myselfMember)
+            }
         }
```
Comment (ktoso) [inline Sources/DistributedActors/ClusterSystem.swift:504]: This probably is enough to use the `cluster.down(node:)` API, so then we don't need the task and async await dance 👍

File: Tests/DistributedActorsTests/TraversalTests.swift [hunk 4]
```diff
@@ -86,6 +86,7 @@ final class TraversalTests: ClusterSystemXCTestCase {
                 // "downingStrategy", "swim", "nodeDeathWatcher", "gossip", "cluster", "leadership",
                 // end of cluster actors ---
                 "clusterEvents",
+                "clusterMembership",
```
Comment (ktoso) [inline Tests/DistributedActorsTests/TraversalTests.swift:89]: I think this one is fine as not-distributed actor; so it won't be in the tree either.

**Review by ktoso — APPROVED** [review]
LGTM! Just very minor comments

File: Sources/DistributedActors/Cluster/ClusterControl.swift [hunk 5]
```diff
@@ -42,37 +43,62 @@ public struct ClusterControl {
...
+            try await self._membershipSnapshotHolder.update(snapshot)
+        }
     }
 
-    private let membershipSnapshotLock: Lock
     private let _membershipSnapshotHolder: MembershipHolder
-    private class MembershipHolder {
+    internal distributed actor MembershipHolder {
```
[the remark this answers is not in the document]
Comment (yim-lee) [reply Sources/DistributedActors/Cluster/ClusterControl.swift:58]: Ah

File: Sources/DistributedActors/ClusterSystem.swift (hunk 3 above)
Comment (yim-lee) [reply Sources/DistributedActors/ClusterSystem.swift]: amended https://github.com/apple/swift-distributed-actors/pull/969/commits/3efd08ffa64e97123a35b9c90d5e9290e700ed87

**Comment by yim-lee** [comment]
Test failure: https://github.com/apple/swift-distributed-actors/issues/970

@swift-server-bot test this please

**Comment by yim-lee** [comment]
Test failure: https://github.com/apple/swift-distributed-actors/issues/971

@swift-server-bot test this please

**Comment by ktoso** [comment]
Thanks for logging the failures, will investigate 👍

**Review by ktoso — APPROVED** [review]
LGTM, thanks!

**Pull request merged**